async/source.go

```diff
@@ -32,59 +56,71 @@ type MessageSource interface {
 // NewMessageSource returns a new insurance asynchronous message source, given
 // an AsyncMessageSource. When Close is called on the AsyncMessageSource,
 // this is also propogated to the underlying AsyncMessageSource.
-func NewMessageSource(ams substrate.AsyncMessageSource) MessageSource {
-	return &messageSourceAdapter{
-		ams,
+func NewMessageSource(source substrate.AsyncMessageSource, opts ...MessageSourceOption) MessageSource {
+	msa := &messageSourceAdapter{
+		source:    source,
+		consumers: 1,
+	}
+
+	for _, opt := range opts {
+		opt(msa)
 	}
+
+	return msa
 }
 
 type messageSourceAdapter struct {
-	ac substrate.AsyncMessageSource
+	source    substrate.AsyncMessageSource
+	msgBuffer uint
+	ackBuffer uint
+	consumers uint
 }
 
 func (a *messageSourceAdapter) ConsumeMessages(ctx context.Context, handler ConsumerMessageHandler) error {
 	rg, ctx := rungroup.New(ctx)
 
-	messages := make(chan substrate.Message)
-	acks := make(chan substrate.Message)
+	messages := make(chan substrate.Message, a.msgBuffer)
+	acks := make(chan substrate.Message, a.ackBuffer)
 
 	rg.Go(func() error {
-		return a.ac.ConsumeMessages(ctx, messages, acks)
+		return a.source.ConsumeMessages(ctx, messages, acks)
 	})
 
-	rg.Go(func() error {
-		for {
-			select {
-			case msg := <-messages:
-				acked := false
-				ackFunc := func() error {
-					if acked {
-						return nil
+	for i := uint(0); i < a.consumers; i++ {
+		rg.Go(func() error {
+			for {
+				select {
+				case msg := <-messages:
+					acked := false
+					ackFunc := func() error {
+						if acked {
+							return nil
+						}
+						select {
+						case acks <- msg:
+							acked = true
+							return nil
+						case <-ctx.Done():
+							return ctx.Err()
+						}
 					}
-					select {
-					case acks <- msg:
-						acked = true
-						return nil
-					case <-ctx.Done():
-						return ctx.Err()
+					if err := handler(ctx, msg, ackFunc); err != nil {
+						return err
 					}
+				case <-ctx.Done():
+					return nil
 				}
-				if err := handler(ctx, msg, ackFunc); err != nil {
-					return err
-				}
-			case <-ctx.Done():
-				return nil
 			}
-		}
-	})
+		})
+	}
 
 	return rg.Wait()
 }
 
 func (a *messageSourceAdapter) Close() error {
-	return a.ac.Close()
+	return a.source.Close()
```

Comment:
I feel like we ought to shutdown goroutines launched in `ConsumeMessages` first, wdyt?